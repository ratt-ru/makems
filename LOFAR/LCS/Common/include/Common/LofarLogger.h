//#  LofarLogger.h: Macro interface to the lofar logging package
//#
//#  Copyright (C) 2004
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: LofarLogger.h 11806 2008-09-28 21:07:32Z loose $

#ifndef LOFAR_COMMON_LOFARLOGGER_H
#define LOFAR_COMMON_LOFARLOGGER_H

// \file
// Macro interface to the lofar logging package.

//# Include the correct set of macros, depending on the availability of the
//# log4cplus or log4cxx packages.
#if defined(HAVE_LOG4CPLUS)
# include <Common/LofarLog4Cplus.h>
#elif defined(HAVE_LOG4CXX)
# include <Common/LofarLog4Cxx.h>
#else
# include <Common/LofarLogCout.h>
#endif

#include <Common/Exceptions.h>

namespace LOFAR 
{

// \ingroup Common
// \addtogroup LofarLogger
  // @{


//# --------------------- Common definitions ---------------
//
// \name Common definitions 
// @{

// The package name of the current package. This variable should be defined in
// the file lofar_config.h which is generated by the lofarconf script. Here,
// we will make sure that LOFARLOGGER_PACKAGE is always defined.
#ifndef LOFARLOGGER_PACKAGE
# define LOFARLOGGER_PACKAGE "Unknown_package"
#endif

// The full package name is defined as the concatenation of the package name
// and an optional sub-package name that may be defined by the user.
#ifdef LOFARLOGGER_SUBPACKAGE
# define LOFARLOGGER_FULLPACKAGE LOFARLOGGER_PACKAGE "." LOFARLOGGER_SUBPACKAGE
#else
# define LOFARLOGGER_FULLPACKAGE LOFARLOGGER_PACKAGE
#endif

// @}

//#------------------------ Assert and FailWhen -------------------------------
//#
//# NB: THROW is defined in package-dependant file.
//
// \name Assert and FailWhen
// \note The \c DBG version of \c ASSERT and \c FAILWHEN will only be in your
// compiled code when the (pre)compiler flag \c ENABLE_DBGASSERT is defined.
// @{

// If the condition of assert is NOT met a logrequest is sent to the logger
// <tt>\<module\>.EXCEPTION</tt> and an AssertError exception is thrown.
#define ASSERT(cond) do { \
	if (!(cond)) THROW(::LOFAR::AssertError, "Assertion: " #cond); \
	} while(0)


// If the condition of assert is NOT met, a logrequest is sent to the logger
// <tt>\<module\>.EXCEPTION</tt> and an AssertError exception is thrown.
#define ASSERTSTR(cond,stream) do { \
	if (!(cond)) THROW(::LOFAR::AssertError, \
			   "Assertion: " #cond "; " << stream); \
	} while(0)

// If the condition of failwhen is met, a logrequest is sent to the logger
// <tt>\<module\>.EXCEPTION</tt> and an AssertError exception is thrown.
#define FAILWHEN(cond) do { \
	if (cond) THROW(::LOFAR::AssertError, "Failtest: " #cond); \
	} while(0)

#ifdef ENABLE_DBGASSERT
#define DBGASSERT(cond)				ASSERT(cond)
#define DBGASSERTSTR(cond,stream)	ASSERTSTR(cond,stream)
#define DBGFAILWHEN(cond)			FAILWHEN(cond)
#else
#define DBGASSERT(cond)
#define DBGASSERTSTR(cond,stream)
#define DBGFAILWHEN(cond)
#endif

// @}
    
} // namespace LOFAR

#endif // file read before
