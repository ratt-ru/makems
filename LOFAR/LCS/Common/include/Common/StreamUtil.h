//#  StreamUtil.h: useful stream manipulation methods.
//#
//#  Copyright (C) 2002-2003
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: StreamUtil.h 9368 2006-10-03 14:02:18Z loose $

#ifndef LOFAR_COMMON_STREAMUTIL_H
#define LOFAR_COMMON_STREAMUTIL_H

// \file
// Useful stream manipulation methods.

//# Includes
#include <Common/LofarTypes.h>
#include <Common/lofar_string.h>
#include <Common/lofar_vector.h>
#include <Common/lofar_iosfwd.h>

namespace LOFAR
{
  // Handles indentation of text lines. Every time an Indent object is
  // constructed, the static member \c level is incremented by one, and every
  // time an Indent object is destructed \c level is decremented by one. To
  // increment the amount of indentation you simply create an Indent
  // object. When this object goes out of scope, the amount of indentation is
  // automagically decremented.
  class Indent {
  public:
    // Constructor. Increments indentation level.
    Indent() { lvl++; }
    // Destructor. Decrements indentation level.
    ~Indent() { lvl--; }
    // Return the amount of indentation.
    static uint level() { return lvl; }
    // Return the token to be printed per indentation level
    static const string& token() { return tok; }
  private:
    // Indentation level.
    static uint lvl;
    // Token to be printed per indentation level.
    static const string tok;
  };


  // Write a vector to an ostream with a given separator, prefix and postfix.
  template<class T>
  void writeVector (std::ostream& os, const std::vector<T>& vec,
		    const char* separator=",",
		    const char* prefix="[", const char* postfix="]")
  {
    os << prefix;
    for (uint i = 0; i < vec.size(); i++) {
      if (i > 0) os << separator;
      os << vec[i];
    }
    os << postfix;
  }


  // Print an indentation that depends on the number of Indent objects
  // currently in existence.
  inline ostream& indent(ostream& os) 
  {
    for (uint i = 0; i < Indent::level(); ++i) {
      os << Indent::token();
    }
    return os;
  }


  // Print the contents of a vector enclosed in square brackets, using a comma
  // as separator.
  // \note operator<<() must be defined for type \c T.
  template<typename T>
  inline ostream& operator<<(ostream& os, const vector<T>& v)
  {
    writeVector(os, v, ",", "[", "]");
    return os;
  }


} // namespace LOFAR

#endif
