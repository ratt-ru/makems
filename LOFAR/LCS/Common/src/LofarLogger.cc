//#  LofarLogger.cc: Interface to the log4cplus logging package
//#
//#  Copyright (C) 2004
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: LofarLogger.cc 11757 2008-09-17 09:49:47Z loose $

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>

//# Includes
#include <cstdio>					// snprintf
#include <unistd.h>					// readlink
#include <libgen.h>				// basename
#include <cstring>
#include <Common/LofarLogger.h>

#if defined(HAVE_LOG4CXX)
# include <log4cxx/ndc.h>
#endif


namespace LOFAR {

//#------------------------- Internal implementation -----------------------------
//
// lofarLoggerInitNode()
//
// Creates a NDC with the text "application@node" and pushes it
// on the NDC stack
void lofarLoggerInitNode(void) {
	const int	MAXLEN = 128;
	int		applNameLen = 0;
	char	hostName [MAXLEN];
	char	applName [MAXLEN];
	char	loggerId [MAXLEN];

	// try to resolve the hostname
	if (gethostname (hostName, MAXLEN-1) < 0) {
		hostName[0]='\0';
	}

	// try to resolve the applicationname
	applNameLen = readlink("/proc/self/exe", applName, MAXLEN-1);
	if (applNameLen >= 0)
		applName[applNameLen] = '\0';
	else
		strcpy (applName, "");
//	}	

	// construct loggerId and register it.
	snprintf(loggerId, MAXLEN-1, "%s@%s", basename(applName), hostName);
#if defined(HAVE_LOG4CPLUS)
	log4cplus::getNDC().push(loggerId);
#elif defined(HAVE_LOG4CXX)
        log4cxx::NDC::push(loggerId);
#endif
}

#if defined(HAVE_LOG4CPLUS)
using namespace log4cplus;

//# ------------------------ implement the five trace levels ------------------------
const LogLevel	TRACE1_LOG_LEVEL	= 1;
const LogLevel	TRACE2_LOG_LEVEL	= 2;
const LogLevel	TRACE3_LOG_LEVEL	= 3;
const LogLevel	TRACE4_LOG_LEVEL	= 4;
const LogLevel	TRACE5_LOG_LEVEL	= 5;
const LogLevel	TRACE6_LOG_LEVEL	= 6;
const LogLevel	TRACE7_LOG_LEVEL	= 7;
const LogLevel	TRACE8_LOG_LEVEL	= 8;

#define _TRACE1_STRING	LOG4CPLUS_TEXT("TRACE1")
#define _TRACE2_STRING	LOG4CPLUS_TEXT("TRACE2")
#define _TRACE3_STRING	LOG4CPLUS_TEXT("TRACE3")
#define _TRACE4_STRING	LOG4CPLUS_TEXT("TRACE4")
#define _TRACE5_STRING	LOG4CPLUS_TEXT("TRACE5")
#define _TRACE6_STRING	LOG4CPLUS_TEXT("TRACE6")
#define _TRACE7_STRING	LOG4CPLUS_TEXT("TRACE7")
#define _TRACE8_STRING	LOG4CPLUS_TEXT("TRACE8")

tstring traceLevel2String(LogLevel ll) {
	switch (ll) {
	case TRACE1_LOG_LEVEL:	return _TRACE1_STRING;
	case TRACE2_LOG_LEVEL:	return _TRACE2_STRING;
	case TRACE3_LOG_LEVEL:	return _TRACE3_STRING;
	case TRACE4_LOG_LEVEL:	return _TRACE4_STRING;
	case TRACE5_LOG_LEVEL:	return _TRACE5_STRING;
	case TRACE6_LOG_LEVEL:	return _TRACE6_STRING;
	case TRACE7_LOG_LEVEL:	return _TRACE7_STRING;
	case TRACE8_LOG_LEVEL:	return _TRACE8_STRING;
	}

	return tstring();		// not found
}

LogLevel string2TraceLevel (const tstring& lname) {
	if (lname == _TRACE1_STRING)	return TRACE1_LOG_LEVEL;
	if (lname == _TRACE2_STRING)	return TRACE2_LOG_LEVEL;
	if (lname == _TRACE3_STRING)	return TRACE3_LOG_LEVEL;
	if (lname == _TRACE4_STRING)	return TRACE4_LOG_LEVEL;
	if (lname == _TRACE5_STRING)	return TRACE5_LOG_LEVEL;
	if (lname == _TRACE6_STRING)	return TRACE6_LOG_LEVEL;
	if (lname == _TRACE7_STRING)	return TRACE7_LOG_LEVEL;
	if (lname == _TRACE8_STRING)	return TRACE8_LOG_LEVEL;

	return NOT_SET_LOG_LEVEL;			// not found
}

LOFAR::LoggerReference theirTraceLoggerRef("TRC");   // create the tracelogger

#elif defined(HAVE_LOG4CXX)

log4cxx::LoggerPtr theirTraceLoggerRef(log4cxx::Logger::getLogger("TRC"));

#endif


// initTracemodule
//
// Function that is used when the TRACE levels are NOT compiled out. It registers
// the TRACEn management routines at the Log4Cplus LogLevelManager and sets up the
// global trace-logger named "TRACE", with the additivity set to false.
// Attached to the trace-logger is one Appender that logs to stderr.
//
void initTraceModule (void) {
#ifdef HAVE_LOG4CPLUS
	//# register our own loglevels
	getLogLevelManager().pushToStringMethod(traceLevel2String);
	getLogLevelManager().pushFromStringMethod(string2TraceLevel);
	
	//# Setup a property object to initialise the TRACE Logger
	helpers::Properties		traceProp;		
	traceProp.setProperty("log4cplus.logger.TRC", "DEBUG, STDERR");
	traceProp.setProperty("log4cplus.additivity.TRC", "false");
	traceProp.setProperty("log4cplus.appender.STDERR", "log4cplus::ConsoleAppender");
	traceProp.setProperty("log4cplus.appender.STDERR.logToStdErr", "true");
	traceProp.setProperty("log4cplus.appender.STDERR.ImmediateFlush", "true");
	traceProp.setProperty("log4cplus.appender.STDERR.layout","log4cplus::PatternLayout");
	traceProp.setProperty("log4cplus.appender.STDERR.layout.ConversionPattern",
							"%D{%y%m%d %H%M%S,%q} [%P] %-6p %c{3} [%F:%L] - %m%n");

	PropertyConfigurator(traceProp).configure();
	Logger::getInstance("TRC").forcedLog(0, "TRACE module activated");

#endif // HAVE_LOG4CPLUS
}

}	// namespace LOFAR
