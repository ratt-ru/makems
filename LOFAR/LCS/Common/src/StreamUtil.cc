//#  StreamUtil.cc: implementation of the stream utilities.
//#
//#  Copyright (C) 2002-2005
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: StreamUtil.cc 9368 2006-10-03 14:02:18Z loose $

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>

#include <Common/StreamUtil.h>

namespace LOFAR
{
  // Initialize static variables.
  uint Indent::lvl = 0;
  const string Indent::tok = ". ";

} // namespace LOFAR
