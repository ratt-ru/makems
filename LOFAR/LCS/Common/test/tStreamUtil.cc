//#  tStreamUtil.cc: test program for the stream utilities
//#
//#  Copyright (C) 2002-2003
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: tStreamUtil.cc 11748 2008-09-16 11:03:51Z loose $

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>

#include <Common/StreamUtil.h>
#include <Common/LofarLogger.h>
#include <Common/lofar_vector.h>
#include <Common/lofar_string.h>
#include <cstring>

using namespace LOFAR;

int main()
{
  INIT_LOGGER("tStreamUtil");
  cout << endl << "Usage of indent:" << endl;
  cout << indent << "Zeroth level, no indentation" << endl;
  {
    Indent id;
    cout << indent << "First level indentation" << endl;
    {
      Indent id;
      cout << indent << "Second level indentation" << endl;
    }
    cout << indent << "Back to first level" << endl;
  }
  cout << indent << "Back to zeroth level" << endl;
  cout << endl;


  vector<int> vi;
  for (int i = 0; i < 20; i++) vi.push_back(i+1);
  cout << "A vector of integers:" << endl
       << vi << endl << endl;
  
  vector<string> vs;
  vs.push_back("aap");   vs.push_back("noot");    vs.push_back("mies");
  vs.push_back("wim");   vs.push_back("zus");     vs.push_back("jet");
  vs.push_back("teun");  vs.push_back("vuur");    vs.push_back("gijs");
  vs.push_back("lam");   vs.push_back("kees");    vs.push_back("bok");
  vs.push_back("weide"); vs.push_back("does");    vs.push_back("hok");
  vs.push_back("duif");  vs.push_back("schapen");
  cout << "A vector of strings:" << endl
       << vs << endl << endl;

  return 0;
}
