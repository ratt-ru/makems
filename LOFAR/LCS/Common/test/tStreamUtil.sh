#!/bin/sh
./runctest.sh tStreamUtil 2>&1 > tStreamUtil.log
