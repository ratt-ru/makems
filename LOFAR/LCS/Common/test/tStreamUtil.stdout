
Usage of indent:
Zeroth level, no indentation
. First level indentation
. . Second level indentation
. Back to first level
Back to zeroth level

A vector of integers:
[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20]

A vector of strings:
[aap,noot,mies,wim,zus,jet,teun,vuur,gijs,lam,kees,bok,weide,does,hok,duif,schapen]

